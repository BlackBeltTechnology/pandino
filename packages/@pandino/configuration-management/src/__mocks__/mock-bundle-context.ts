import {
  Bundle,
  BundleContext,
  BundleListener,
  BundleManifestHeaders,
  BundleState,
  BundleTracker,
  BundleTrackerCustomizer,
  FrameworkListener,
  ServiceEventType,
  ServiceListener,
  ServiceObjects,
  ServiceProperties,
  ServiceReference,
  ServiceRegistration,
  ServiceTracker,
  ServiceTrackerCustomizer,
} from '@pandino/pandino-api';

export class MockBundleContext implements BundleContext {
  private properties: Record<any, any> = {};
  private serviceListeners: ServiceListener[] = [];
  private readonly registrations: ServiceRegistration<any>[] = [];
  private readonly refMap: Map<string[] | string, ServiceReference<any>> = new Map<
    string[] | string,
    ServiceReference<any>
  >();
  private readonly serviceMap: Map<ServiceReference<any>, any> = new Map<ServiceReference<any>, any>();
  private bundle: Bundle;

  setProperty(name: string, value: string): void {
    if (value === null || value === undefined) {
      delete this.properties[name];
    } else {
      this.properties[name] = value;
    }
  }

  setBundle(bundle: Bundle): void {
    this.bundle = bundle;
  }

  addBundleListener(listener: BundleListener): void {}

  addFrameworkListener(listener: FrameworkListener): void {}

  addServiceListener(listener: ServiceListener, filter?: string): void {
    this.serviceListeners.push(listener);
  }

  equals(other: any): boolean {
    return false;
  }

  getBundle(id?: number): Bundle;
  getBundle(): Bundle;
  getBundle(id?: number): Bundle {
    return this.bundle;
  }

  getBundles(): Bundle[] {
    return [];
  }

  getProperty(key: string): string {
    return '';
  }

  getService<S>(reference: ServiceReference<S>): S {
    return this.serviceMap.get(reference);
  }

  getServiceReference<S>(identifier: string): ServiceReference<S> {
    return this.refMap.get(identifier);
  }

  getServiceReferences<S>(identifier: string, filter?: string): ServiceReference<S>[] {
    return [];
  }

  installBundle(locationOrHeaders: string | BundleManifestHeaders): Promise<Bundle> {
    return Promise.resolve(undefined);
  }

  registerService<S>(
    identifiers: string[] | string,
    service: S,
    properties: ServiceProperties = {},
  ): ServiceRegistration<S> {
    const ref: ServiceReference<any> = {
      getBundle: () => this.getBundle(),
      getProperties: () => properties,
      getProperty: (key: string) => properties[key],
      compareTo: () => 0,
      getPropertyKeys: () => Object.keys(properties),
      getUsingBundles: () => [],
      isAssignableTo(bundle: Bundle, className: string): boolean {
        return true;
      },
      hasObjectClass: (objectClass: string) => false,
    };
    this.refMap.set(identifiers, ref);
    const reg: ServiceRegistration<any> = {
      getReference: () => ref,
      setProperties: () => undefined,
      unregister: () => undefined,
      getProperties: () => properties,
      getProperty: (key: string) => properties[key],
      getPropertyKeys: () => Object.keys(properties),
    };
    this.serviceMap.set(ref, service);
    this.registrations.push(reg);
    this.serviceListeners.forEach((listener) => {
      listener.serviceChanged({
        getType(): ServiceEventType {
          return 'REGISTERED';
        },
        getServiceReference: () => ref,
      });
    });
    return reg;
  }

  removeBundleListener(listener: BundleListener): void {}

  removeFrameworkListener(listener: FrameworkListener): void {}

  removeServiceListener(listener: ServiceListener): void {}

  ungetService<S>(reference: ServiceReference<S>): boolean {
    return false;
  }

  getAllServiceReferences(identifier: string, filter: string): Array<ServiceReference<any>> {
    return [];
  }

  getServiceObjects<S>(reference: ServiceReference<S>): ServiceObjects<S> | undefined {
    return undefined;
  }

  trackBundle<T>(trackedStates: BundleState[], customizer: Partial<BundleTrackerCustomizer<T>>): BundleTracker<T> {
    return undefined;
  }

  trackService<S, T>(
    identifierOrFilter: string,
    customizer: Partial<ServiceTrackerCustomizer<S, T>>,
  ): ServiceTracker<S, T> {
    return undefined;
  }
}
