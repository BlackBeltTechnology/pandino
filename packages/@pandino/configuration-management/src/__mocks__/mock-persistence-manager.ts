import type { PersistenceManager } from '@pandino/persistence-manager-api';
import type { ServiceProperties } from '@pandino/pandino-api';

export class MockPersistenceManager implements PersistenceManager {
  private readonly map: Map<string, ServiceProperties> = new Map<string, ServiceProperties>();

  constructor(json: string) {
    const parsed = JSON.parse(json || '{}');
    Object.keys(parsed).forEach((key: string) => {
      this.map.set(key, parsed[key]);
    });
  }

  delete(pid: string): void {
    this.map.delete(pid);
  }

  exists(pid: string): boolean {
    return this.map.has(pid);
  }

  getProperties(): ServiceProperties[] {
    return Array.from(this.map.values());
  }

  load(pid: string): ServiceProperties | undefined {
    return this.map.get(pid);
  }

  store(pid: string, properties: ServiceProperties): void {
    this.map.set(pid, properties);
  }

  dump(): string {
    let output: any = {};
    this.map.forEach((value: ServiceProperties, key: string) => {
      output[key] = value;
    });
    return JSON.stringify(output, null, 4);
  }
}
