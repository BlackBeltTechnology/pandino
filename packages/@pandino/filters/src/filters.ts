import type { FilterEvaluator, FilterNode, FilterOperator, SemVerOperator } from './types';
import { FilterOperatorSymbolMapping } from './types';

export const convert = (attrs: Record<string, any>): FilterNode => {
  const filters: FilterNode[] = [];

  for (let [_, value] of Object.entries(attrs)) {
    filters.push(parseFilter(value.toString()));
  }

  let filter: FilterNode;

  if (filters.length === 1) {
    filter = filters[0];
  } else if (Object.keys(attrs).length > 1) {
    filter = { attribute: undefined, operator: 'and', children: filters };
  } else if (filters.length === 0) {
    filter = { attribute: undefined, operator: 'eq', value: '*' };
  }

  return filter!;
};

export const evaluateFilter: FilterEvaluator = (data: any, query: string): boolean => {
  const node = parseFilter(query);
  return evaluateFilterNode(node, data);
};

export function evaluateFilterNode(node: FilterNode, data: any): boolean {
  if (!node.operator && node.children) {
    return evaluateFilterNode(node.children[0], data);
  }

  switch (node.operator) {
    case 'and':
      return (
        node.children?.every(function (child) {
          return evaluateFilterNode(child, data);
        }) ?? false
      );
    case 'or':
      return (
        node.children?.some(function (child) {
          return evaluateFilterNode(child, data);
        }) ?? false
      );
    case 'not':
      return node.children ? !evaluateFilterNode(node.children[0], data) : false;
    default:
      return evaluateComparison(node, data);
  }
}

export function serializeFilter(node: FilterNode): string | undefined {
  if (node.operator && node.attribute && node.value !== undefined) {
    return `(${node.attribute}${FilterOperatorSymbolMapping[node.operator]}${node.value})`;
  }

  if (node.operator && node.children && node.children.length > 0) {
    const childFilters = node.children.map(serializeFilter).join('');
    return `(${FilterOperatorSymbolMapping[node.operator]}${childFilters})`;
  }

  if (node.expression) {
    return `(${node.expression})`;
  }

  return undefined;
}

export function parseFilter(filter: string): FilterNode | never {
  const filtered = filter.replace(/\s/g, '');
  const stack = [];
  let current: FilterNode = {};

  for (let i = 0; i < filtered.length; i++) {
    const char = filtered.charAt(i);

    if (char === '(') {
      const node: FilterNode = {};
      current.children = current.children || [];
      current.children.push(node);
      stack.push(current);
      current = node;
    } else if (char === ')') {
      if (!stack.length) {
        throw new Error('Unmatched closing parenthesis');
      }
      // @ts-ignore
      current = stack.pop();
    } else if (char === '|') {
      current.operator = 'or';
    } else if (char === '&') {
      current.operator = 'and';
    } else if (char === '!') {
      current.operator = 'not';
    } else {
      if (!current.expression) {
        current.expression = '';
      }
      current.expression += char;
    }
  }

  splitNodeExpression(current);

  if (
    current &&
    current.operator === undefined &&
    current.value === undefined &&
    current.expression === undefined &&
    current.children &&
    current.children.length === 1
  ) {
    return current.children[0];
  }

  return current;
}

function evaluateComparison(comparison: FilterNode, data: any): boolean {
  let attribute = comparison.attribute;
  let operator = comparison.operator;
  let value = comparison.value;

  // Traverse the nested attributes to get the actual value
  // const attributePath = attribute ? attribute.split('.') : [];
  let current = attribute ? data[attribute] : undefined;
  // while (attributePath.length > 0 && current) {
  //   current = current[attributePath.shift()!];
  // }

  // Handle missing attributes gracefully
  if (current === undefined) {
    return false;
  }

  let typedValue = value;
  if (current !== null) {
    if (typeof current === 'number') {
      typedValue = Number(value);
    } else if (typeof current === 'boolean') {
      typedValue = value === 'true';
    } else if (isSemVer(current)) {
      return evaluateSemver(current, operator as SemVerOperator, value);
    }
  }

  const arrayIncludes = operator === 'eq' && Array.isArray(current);
  const present = typedValue === '*';
  const leadStar = typeof typedValue === 'string' && typedValue.startsWith('*');
  const trailStar = typeof typedValue === 'string' && typedValue.endsWith('*');
  const startsWith = !leadStar && trailStar;
  const endsWith = leadStar && !trailStar;
  const contains = leadStar && trailStar;

  switch (operator) {
    case 'eq':
      if (arrayIncludes) {
        return (current as Array<any>).includes(typedValue);
      } else if (present) {
        return current !== null && current !== undefined;
      } else if (contains) {
        const mut = typedValue.substring(1, typedValue.length - 2);
        return current.includes(mut);
      } else if (startsWith) {
        const mut = typedValue.substring(0, typedValue.length - 2);
        return current.startsWith(mut);
      } else if (endsWith) {
        const mut = typedValue.substring(1);
        return current.endsWith(mut);
      }
      return current === typedValue;
    case 'gt':
      return current > typedValue;
    case 'gte':
      return current >= typedValue;
    case 'lt':
      return current < typedValue;
    case 'lte':
      return current <= typedValue;
    default:
      return false;
  }
}

function splitNodeExpression(node: FilterNode): void {
  const kv: Record<string, FilterOperator> = {
    '<=': 'lte',
    '>=': 'gte',
    '=': 'eq',
    '<': 'lt',
    '>': 'gt',
  };

  if (node.expression) {
    let trimmed = node.expression.trim();
    for (const key in kv) {
      if (trimmed.includes(key)) {
        const split = trimmed.split(key);

        node.operator = kv[key];
        node.attribute = split[0].trim();
        node.value = split[1].trim();

        delete node.expression;
        break;
      }
    }
  }

  if (Array.isArray(node.children)) {
    node.children.forEach(splitNodeExpression);
  }
}

export function isSemVer(version: string): boolean {
  const semverRegex =
    /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/;
  return semverRegex.test(version);
}

export function evaluateSemver(version: string, operator: SemVerOperator, targetVersion: string) {
  // Split the version strings into major, minor, and patch components
  const [vMajor, vMinor, vPatch] = version.split('.').map(Number);
  const [tMajor, tMinor, tPatch] = targetVersion.split('.').map(Number);

  // Evaluate the comparison based on the specified operator
  switch (operator) {
    case 'lte':
      return vMajor < tMajor ? true : vMajor > tMajor ? false : vMinor < tMinor ? true : vMinor > tMinor ? false : vPatch <= tPatch;
    case 'lt':
      return vMajor < tMajor ? true : vMajor > tMajor ? false : vMinor < tMinor ? true : vMinor > tMinor ? false : vPatch < tPatch;
    case 'gte':
      return vMajor > tMajor ? true : vMajor < tMajor ? false : vMinor > tMinor ? true : vMinor < tMinor ? false : vPatch >= tPatch;
    case 'gt':
      return vMajor > tMajor ? true : vMajor < tMajor ? false : vMinor > tMinor ? true : vMinor < tMinor ? false : vPatch > tPatch;
    case 'eq':
      return vMajor === tMajor && vMinor === tMinor && vPatch === tPatch;
    default:
      throw new Error(`Invalid operator: ${operator}`);
  }
}
