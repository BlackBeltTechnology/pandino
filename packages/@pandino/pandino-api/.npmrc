engine-strict=true
scope=pandino
