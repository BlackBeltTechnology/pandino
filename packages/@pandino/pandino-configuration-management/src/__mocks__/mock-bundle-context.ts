import {
  Bundle,
  BundleContext,
  BundleListener,
  BundleManifestHeaders,
  FilterApi,
  FrameworkListener,
  ServiceEventType,
  ServiceListener,
  ServiceProperties,
  ServiceReference,
  ServiceRegistration,
} from '@pandino/pandino-api';

export class MockBundleContext implements BundleContext {
  private properties: Record<any, any> = {};
  private serviceListeners: ServiceListener[] = [];
  private readonly registrations: ServiceRegistration<any>[] = [];
  private readonly refMap: Map<string[] | string, ServiceReference<any>> = new Map<
    string[] | string,
    ServiceReference<any>
  >();
  private readonly serviceMap: Map<ServiceReference<any>, any> = new Map<ServiceReference<any>, any>();
  private bundle: Bundle;

  setProperty(name: string, value: string): void {
    if (value === null || value === undefined) {
      delete this.properties[name];
    } else {
      this.properties[name] = value;
    }
  }

  setBundle(bundle: Bundle): void {
    this.bundle = bundle;
  }

  addBundleListener(listener: BundleListener): void {}

  addFrameworkListener(listener: FrameworkListener): void {}

  addServiceListener(listener: ServiceListener, filter?: string): void {
    this.serviceListeners.push(listener);
  }

  createFilter(filter: string): FilterApi {
    return undefined;
  }

  equals(other: any): boolean {
    return false;
  }

  getBundle(id?: number): Bundle;
  getBundle(): Bundle;
  getBundle(id?: number): Bundle {
    return this.bundle;
  }

  getBundles(): Bundle[] {
    return [];
  }

  getProperty(key: string): string {
    return '';
  }

  getService<S>(reference: ServiceReference<S>): S {
    return this.serviceMap.get(reference);
  }

  getServiceReference<S>(identifier: string): ServiceReference<S> {
    return this.refMap.get(identifier);
  }

  getServiceReferences<S>(identifier: string, filter?: string): ServiceReference<S>[] {
    return [];
  }

  installBundle(locationOrHeaders: string | BundleManifestHeaders): Promise<Bundle> {
    return Promise.resolve(undefined);
  }

  registerService<S>(
    identifiers: string[] | string,
    service: S,
    properties: ServiceProperties = {},
  ): ServiceRegistration<S> {
    const ref: ServiceReference<any> = {
      getBundle: () => this.getBundle(),
      getProperties: () => properties,
      getProperty: (key: string) => properties[key],
      compareTo: () => 0,
      getPropertyKeys: () => Object.keys(properties),
      getUsingBundles: () => [],
      isAssignableTo(bundle: Bundle, className: string): boolean {
        return true;
      },
      hasObjectClass: (objectClass: string) => false,
    };
    this.refMap.set(identifiers, ref);
    const reg: ServiceRegistration<any> = {
      getReference: () => ref,
      setProperties: () => undefined,
      unregister: () => undefined,
      getProperties: () => properties,
      getProperty: (key: string) => properties[key],
      getPropertyKeys: () => Object.keys(properties),
    };
    this.serviceMap.set(ref, service);
    this.registrations.push(reg);
    this.serviceListeners.forEach((listener) => {
      listener.serviceChanged({
        getType(): ServiceEventType {
          return 'REGISTERED';
        },
        getServiceReference: () => ref,
      });
    });
    return reg;
  }

  removeBundleListener(listener: BundleListener): void {}

  removeFrameworkListener(listener: FrameworkListener): void {}

  removeServiceListener(listener: ServiceListener): void {}

  ungetService<S>(reference: ServiceReference<S>): boolean {
    return false;
  }

  getAllServiceReferences(identifier: string, filter: string): Array<ServiceReference<any>> {
    return [];
  }
}
