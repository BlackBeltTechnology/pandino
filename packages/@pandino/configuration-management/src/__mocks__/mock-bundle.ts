import type { Bundle, BundleContext, BundleManifestHeaders, BundleState, ServiceReference } from '@pandino/pandino-api';
import { MockBundleContext } from './mock-bundle-context';

export class MockBundle implements Bundle {
  private readonly context: BundleContext;
  private readonly location: string;
  private readonly symbolicName: string;
  private readonly version: string;

  constructor(context: MockBundleContext, location: string, symbolicName: string, version: string) {
    this.context = context;
    this.location = location;
    this.symbolicName = symbolicName;
    this.version = version;
    context.setBundle(this);
  }

  getBundleContext(): BundleContext {
    return this.context;
  }

  getBundleId(): number {
    return 0;
  }

  getHeaders(): BundleManifestHeaders {
    return undefined as any;
  }

  getLocation(): string {
    return this.location;
  }

  getRegisteredServices(): ServiceReference<any>[] {
    return [];
  }

  getServicesInUse(): ServiceReference<any>[] {
    return [];
  }

  getState(): BundleState {
    return undefined as any;
  }

  getSymbolicName(): string {
    return this.symbolicName;
  }

  getUniqueIdentifier(): string {
    return '';
  }

  getVersion(): string {
    return this.version;
  }

  start(options?: BundleState): Promise<void> {
    return Promise.resolve(undefined);
  }

  stop(options?: BundleState): Promise<void> {
    return Promise.resolve(undefined);
  }

  uninstall(): Promise<void> {
    return Promise.resolve(undefined);
  }

  update(headers: BundleManifestHeaders, bundle?: Bundle): Promise<void> {
    return Promise.resolve(undefined);
  }
}
