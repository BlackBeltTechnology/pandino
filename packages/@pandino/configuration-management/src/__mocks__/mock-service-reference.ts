import type { Bundle, ServiceProperties, ServiceReference } from '@pandino/pandino-api';
import { SERVICE_PID } from '@pandino/pandino-api';

export class MockServiceReference implements ServiceReference<any> {
  private readonly bundle: Bundle;
  private readonly pids: any;

  constructor(bundle: Bundle, pids: any) {
    this.bundle = bundle;
    this.pids = pids;
  }

  hasObjectClass(objectClass: string): boolean {
    return false;
  }

  compareTo(other: ServiceReference<any>): number {
    return 0;
  }

  getBundle(): Bundle {
    return this.bundle;
  }

  getProperties(): ServiceProperties {
    return undefined as any;
  }

  getProperty(key: string): any {
    if (SERVICE_PID === key) {
      return this.pids;
    }
  }

  getPropertyKeys(): Array<string> {
    return [];
  }

  getUsingBundles(): Bundle[] {
    return [];
  }

  isAssignableTo(bundle: Bundle, className: string): boolean {
    return false;
  }
}
