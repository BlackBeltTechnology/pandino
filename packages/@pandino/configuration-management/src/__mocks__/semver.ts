import { SemVer } from '@pandino/pandino-api';

export function createVersionMock(version: string): any {
  return {
    __version: version,
    toString: () => version,
    compare: (other: SemVer) => {
      return other.toString() === version ? 0 : -1; // this is a super simplified mock, please adjust when needed
    },
  };
}
