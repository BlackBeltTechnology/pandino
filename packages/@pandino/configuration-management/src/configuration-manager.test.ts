import { describe, beforeEach, expect, it, vi } from 'vitest';
import { SERVICE_PID } from '@pandino/pandino-api';
import type { Bundle, BundleContext, Logger, ServiceEvent, ServiceReference } from '@pandino/pandino-api';
import { evaluateFilter } from '@pandino/filters';
import type { ManagedService } from '@pandino/configuration-management-api';
import { MockBundleContext } from './__mocks__/mock-bundle-context';
import { MockBundle } from './__mocks__/mock-bundle';
import { MockPersistenceManager } from './__mocks__/mock-persistence-manager';
import { ConfigurationManager } from './configuration-manager';

describe('ConfigurationManager', function () {
  let persistenceManager: MockPersistenceManager;
  let context: BundleContext;
  let bundle: Bundle;
  let cm: ConfigurationManager;
  let mockDebug = vi.fn();
  let logger: Logger = {
    debug: mockDebug,
  } as unknown as Logger;

  beforeEach(() => {
    mockDebug.mockClear();
    context = new MockBundleContext();
    bundle = new MockBundle(context as MockBundleContext, 'test.bundle.location', '@test/my-bundle', '0.0.0');
    persistenceManager = new MockPersistenceManager(`{
      "my.component.pid": {
        "service.pid": "my.component.pid",
        "port" : 300
      }
    }`);
    cm = new ConfigurationManager(context, logger, evaluateFilter, persistenceManager);
  });

  it('listConfigurations()', () => {
    persistenceManager = new MockPersistenceManager(`{
      "my.component.pid": {
        "service.pid": "my.component.pid",
        "port" : 300, 
        "collection" : [2, 3, 4],  
        "complex": { 
          "a" : 1, 
          "b" : "two"
        }
      },
      "my.other.pid": {
        "service.pid": "my.other.pid",
        "key": "value"
      }
    }`);
    cm = new ConfigurationManager(context, logger, evaluateFilter, persistenceManager);

    expect(cm.listConfigurations().length).toEqual(2);

    const [config1, config2] = cm.listConfigurations();

    expect(config1.getPid()).toEqual('my.component.pid');
    expect(config1.getProperties()).toEqual({
      collection: [2, 3, 4],
      complex: { a: 1, b: 'two' },
      port: 300,
      'service.pid': 'my.component.pid',
    });

    expect(config2.getPid()).toEqual('my.other.pid');
    expect(config2.getProperties()).toEqual({ key: 'value', 'service.pid': 'my.other.pid' });
  });

  it('listConfigurations() with filter', () => {
    persistenceManager = new MockPersistenceManager(`{
      "my.component.pid": {
        "service.pid": "my.component.pid",
        "port" : 300, 
        "collection" : [2, 3, 4],  
        "complex": { 
          "a" : 1, 
          "b" : "two"
        }
      },
      "my.other.pid": {
        "service.pid": "my.other.pid",
        "key": "value"
      }
    }`);
    cm = new ConfigurationManager(context, logger, evaluateFilter, persistenceManager);
    const configurations = cm.listConfigurations('(key=value)');

    expect(configurations.length).toEqual(1);

    const [config2] = configurations;

    expect(config2.getPid()).toEqual('my.other.pid');
    expect(config2.getProperties()).toEqual({ key: 'value', 'service.pid': 'my.other.pid' });
  });

  it('getConfiguration()', () => {
    cm = new ConfigurationManager(context, logger, evaluateFilter, persistenceManager);

    const config = cm.getConfiguration('my.component.pid');

    expect(config.getPid()).toEqual('my.component.pid');
    expect(config.getBundleLocation()).toEqual('test.bundle.location');
    expect(persistenceManager.dump()).toMatchSnapshot();
  });

  it('configuration update()', () => {
    cm = new ConfigurationManager(context, logger, evaluateFilter, persistenceManager);

    const config = cm.getConfiguration('my.component.pid');

    expect(persistenceManager.dump()).toMatchSnapshot();

    config.update({
      ...config.getProperties(),
      'new-prop': 'new-value',
    });

    expect(persistenceManager.dump()).toMatchSnapshot();
  });

  it('configuration delete()', () => {
    cm = new ConfigurationManager(context, logger, evaluateFilter, persistenceManager);

    const config = cm.getConfiguration('my.component.pid');

    expect(persistenceManager.dump()).toMatchSnapshot();

    config.delete();

    expect(() => config.getPid()).toThrow();
    expect(persistenceManager.dump()).toMatchSnapshot();
  });

  it('serviceChanged for configuration event', () => {
    cm = new ConfigurationManager(context, logger, evaluateFilter, persistenceManager);

    expect((cm as any).eventListeners.size).toEqual(0);

    const mockService: any = {
      configurationEvent: vi.fn(),
    };
    (context as any).getService = () => mockService;
    const mockServiceReference: any = {
      getProperty: (prop: string) => {
        if (prop === SERVICE_PID) {
          return 'mock.pid';
        }
      },
    };
    const registeredEvent: ServiceEvent = {
      getType: () => 'REGISTERED',
      getServiceReference: () => mockServiceReference,
    };

    cm.serviceChanged(registeredEvent);

    expect((cm as any).eventListeners.size).toEqual(1);
    expect((cm as any).eventListeners.has('mock.pid')).toEqual(true);
    expect((cm as any).eventListeners.get('mock.pid').length).toEqual(1);
    expect((cm as any).eventListeners.get('mock.pid')[0]).toEqual(mockService);

    const unregisteringEvent: ServiceEvent = {
      getType: () => 'UNREGISTERING',
      getServiceReference: () => mockServiceReference,
    };

    cm.serviceChanged(unregisteringEvent);

    expect((cm as any).eventListeners.size).toEqual(1);
    expect((cm as any).eventListeners.has('mock.pid')).toEqual(true);
    expect((cm as any).eventListeners.get('mock.pid').length).toEqual(0);
  });

  it('serviceChanged for managed service event', () => {
    cm = new ConfigurationManager(context, logger, evaluateFilter, persistenceManager);

    expect((cm as any).managedReferences.size).toEqual(0);

    const mockService: any = {
      updated: vi.fn(),
    };
    (context as any).getService = () => mockService;
    const mockServiceReference: any = {
      getProperty: (prop: string) => {
        if (prop === SERVICE_PID) {
          return 'mock.pid';
        }
      },
      getBundle: () => ({
        getLocation: () => 'test.location.yayy',
      }),
    };
    const registeredEvent: ServiceEvent = {
      getType: () => 'REGISTERED',
      getServiceReference: () => mockServiceReference,
    };

    cm.serviceChanged(registeredEvent);

    expect((cm as any).managedReferences.size).toEqual(1);
    expect((cm as any).managedReferences.has('mock.pid')).toEqual(true);
    expect((cm as any).managedReferences.get('mock.pid').length).toEqual(1);
    expect((cm as any).managedReferences.get('mock.pid')[0]).toEqual(mockServiceReference);

    const unregisteringEvent: ServiceEvent = {
      getType: () => 'UNREGISTERING',
      getServiceReference: () => mockServiceReference,
    };

    cm.serviceChanged(unregisteringEvent);

    expect((cm as any).managedReferences.size).toEqual(1);
    expect((cm as any).managedReferences.has('mock.pid')).toEqual(true);
    expect((cm as any).managedReferences.get('mock.pid').length).toEqual(0);
  });

  describe('initReferencesAddedBeforeManagerActivation', () => {
    it('some references had pre-stored configurations, while others did not', () => {
      persistenceManager = new MockPersistenceManager(`{
        "my.component.pid": {
          "service.pid": "my.component.pid",
          "port" : 300
        }
      }`);
      const refWithInitializedConfig: any = {
        getProperty: (prop: string) => (prop === SERVICE_PID ? 'my.component.pid' : undefined),
        getProperties: () => ({
          [SERVICE_PID]: 'my.component.pid',
          port: 150,
        }),
        getBundle: () => ({
          getLocation: () => 'some.location',
          getSymbolicName: () => '@scope/component-1',
          getVersion: () => '1.2.3',
        }),
      };
      const refWithOwnConfig: any = {
        getProperty: (prop: string) => (prop === SERVICE_PID ? 'my.other-component.pid' : undefined),
        getProperties: () => ({
          [SERVICE_PID]: 'my.other-component.pid',
          isPowerful: true,
        }),
        getBundle: () => ({
          getLocation: () => 'some.other.location',
          getSymbolicName: () => '@scope/component-2',
          getVersion: () => '1.0.0',
        }),
      };
      const refWithoutAnyConfig: any = {
        getProperty: () => {},
        getProperties: () => ({
          isPowerful: false,
        }),
        getBundle: () => ({
          getLocation: () => 'yet.other.location',
          getSymbolicName: () => '@scope/component-3',
          getVersion: () => '3.2.1',
        }),
      };

      const allRefs: any[] = [refWithInitializedConfig, refWithOwnConfig, refWithoutAnyConfig];
      (context as any).getServiceReferences = () => allRefs;

      const serviceForRefWithInitializedConfig: ManagedService = {
        updated: vi.fn(),
      };
      const serviceForRefWithOwnConfig: ManagedService = {
        updated: vi.fn(),
      };
      const serviceToSkip: ManagedService = {
        updated: vi.fn(),
      };

      const allServices: Map<any, ManagedService> = new Map<any, ManagedService>([
        [refWithInitializedConfig, serviceForRefWithInitializedConfig],
        [refWithOwnConfig, serviceForRefWithOwnConfig],
        [refWithoutAnyConfig, serviceToSkip],
      ]);
      (context as any).getService = (ref: any) => allServices.get(ref);
      cm = new ConfigurationManager(context, logger, evaluateFilter, persistenceManager);

      expect((cm as any).managedReferences.size).toEqual(0);

      cm.initReferencesAddedBeforeManagerActivation();

      const managedRefs: Map<string, Array<ServiceReference<ManagedService>>> = (cm as any).managedReferences;

      expect(managedRefs.size).toEqual(2);
      expect(managedRefs.get('my.component.pid')).toEqual([refWithInitializedConfig]);
      expect(managedRefs.get('my.other-component.pid')).toEqual([refWithOwnConfig]);

      expect(serviceForRefWithInitializedConfig.updated).toHaveBeenCalledTimes(1);
      expect(serviceForRefWithInitializedConfig.updated).toHaveBeenCalledWith({
        port: 300,
        [SERVICE_PID]: 'my.component.pid',
      });
      expect(serviceForRefWithOwnConfig.updated).toHaveBeenCalledTimes(1);
      expect(serviceForRefWithOwnConfig.updated).toHaveBeenCalledWith(undefined);
      expect(serviceToSkip.updated).toHaveBeenCalledTimes(0);

      expect(mockDebug).toHaveBeenCalledTimes(1);
      expect(mockDebug).toHaveBeenCalledWith('Updating non-configured Service for PID: my.other-component.pid');
    });
  });
});
