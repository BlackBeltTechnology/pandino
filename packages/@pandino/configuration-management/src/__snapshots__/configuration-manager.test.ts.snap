// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConfigurationManager > configuration delete() 1`] = `
"{
    "my.component.pid": {
        "service.pid": "my.component.pid",
        "port": 300
    }
}"
`;

exports[`ConfigurationManager > configuration delete() 2`] = `"{}"`;

exports[`ConfigurationManager > configuration update() 1`] = `
"{
    "my.component.pid": {
        "service.pid": "my.component.pid",
        "port": 300
    }
}"
`;

exports[`ConfigurationManager > configuration update() 2`] = `
"{
    "my.component.pid": {
        "service.pid": "my.component.pid",
        "port": 300,
        "new-prop": "new-value"
    }
}"
`;

exports[`ConfigurationManager > getConfiguration() 1`] = `
"{
    "my.component.pid": {
        "service.pid": "my.component.pid",
        "port": 300
    }
}"
`;
